src/librustc_mir/interpret/eval_context.rs

```diff
@@ -36,6 +36,11 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
     /// Bounds in scope for polymorphic evaluations.
     pub(crate) param_env: ty::ParamEnv<'tcx>,
 
+    /// Base substitutions for when there are no frames that we can grab them from.
+    // HACK(oli-obk): this is because we don't want to push stack frames for `const_field` and
+    // `const_variant_index`, because that would be expensive
```

Comment:
This should explain when these are used vs. when the frame is used.

Or maybe "get the current frame's subst" should be a machine method, and then the machine used for those two queries could be different than the normal one that uses the top frame.

Reply:
https://github.com/rust-lang/rust/blob/38a92102140b47c736f09cc690e7615fd6776f2e/src/librustc_mir/interpret/eval_context.rs#L372 is not able to figure out the substs of a specific frame id from the stack because it has a `&Frame` as an argument. That `&Frame` isn't even on the stack yet. I'm not sure how to improve this. It's literally what the docs say. no frames -> substs, any frames -> something else.

Reply:
But if `layout_of_local` has an `&Frame`, why can't it use the subst from that frame...?